### PR TITLE
Add a MoL-based WENO code.

### PR DESCRIPTION
**Pull request by IanHawke**
Uses flux differencing: see references in __init__.py.

This is very much a WIP, especially as I've taken code from elsewhere (https://github.com/IanHawke/toy-conslaw, which is itself a poorly documented port of older code) and mashed it into pyro without matching conventions.

The crucial code is the addition of the (low order) WENO operators to mesh.reconstruction and the fvs routine in fluxes.py and its usage in that file.

### CODE REVIEW

**Comment by IanHawke** [comment]
I fiddled a bit more in the branch https://github.com/IanHawke/pyro2/tree/weno_ho which adds the next two schemes (up to k=5, order 9). This requires more ghostzones so is a bit messier: let me know if you'd rather have that version.

**Comment by zingale** [comment]
thanks.  I'll take a look at this over the weekend.

**Comment by zingale** [comment]
this looks good to me.  My one suggestion would be that since there is nothing different about the problem initialization for this solver as compared to `advection`, that we make `advection_weno/problems/` a softlink to `advection/problems/`.  This is what I do, e.g., for `advection_rk/problems/`

File: mesh/reconstruction.py [hunk 1]
```diff
@@ -142,4 +142,89 @@ def flatten_multid(myg, q, xi_x, xi_y, ivars):
...
+# Constants for the WENO reconstruction
+# NOTE: integer division laziness means this WILL fail on python2
+C_3 = np.array([1, 2]) / 3
+a_3 = np.array([[3, -1], [1, 1]]) / 2
+sigma_3 = np.array([[[1, 0], [-2, 1]], [[1, 0], [-2, 1]]])
+
+C_5 = np.array([1, 6, 3]) / 10
+a_5 = np.array([[11, -7, 2], [2, 5, -1], [-1, 5, 2]]) / 6
+sigma_5 = np.array([[[40, 0, 0],
+                        [-124, 100, 0],
```
Comment (zingale) [inline mesh/reconstruction.py]: can you fix the indentation here so the innermost `[` line up?

**Comment by IanHawke** [comment]
Those commits should fix above comments.

In general I'd expect the higher order WENO schemes to need a lower CFL which might make the inputs files in the problem setup not so useful in these cases, but for some reason (surprised me...) it works ok this time.

**Comment by zingale** [comment]
FYI, I'll review the PR on the notes shortly

**Pull request merged**